# Introduction

This project focuses on analyzing the **Spaceship Titanic Dataset** to develop a *machine learning model* that can predict which passengers were transported to an alternate dimension during a collision with a spacetime anomaly. Set in the year 2912, this futuristic scenario presents a unique challenge in the field of data science and interstellar travel safety.

The dataset contains various attributes about the passengers including:

* Demographics
   * Age
   * Name
* Travel Information
   * Home Planet
   * Destination
   * Cabin location
* Onboard Activities
   * RoomService
   * FoodCourt
   * ShoppingMall
   * Spa
   * VRDeck
* Special Statuses
   * VIP
   * CryoSleep

The target variable is whether the passenger was `Transported` to an alternate dimension.

As data scientists working with **Interstellar Rescue Operations**, our objective is to thoroughly explore this data to uncover patterns and insights that can inform the development of a robust predictive model. This will enable rescue crews to identify which passengers were transported and focus their retrieval efforts accordingly.

The analysis will progress through the following key steps:

1. **Exploratory Data Analysis** - Examining the distributions, ranges, and relationships between the features and target variable through statistical summaries and visualizations. Checking data quality and identifying any anomalies in the spaceship's records.

2. **Statistical Inference** - Formulating and testing hypotheses about transportation risk factors and quantifying uncertainty through confidence intervals.

3. **Machine Learning Modeling** - Applying a range of classification algorithms including random forests, gradient boosting, and potentially neural networks to predict transportation likelihood. Tuning hyperparameters and building ensembles to optimize predictive performance.

4. **Model Deployment** - Selecting the top performing model and preparing it for deployment to enable efficient identification of transported passengers in future space anomaly encounters.

Throughout this notebook, detailed commentary will be provided on the analytical approach, key findings, model results and ideas for further enhancement. The goal is to demonstrate a thoughtful, thorough analysis while documenting reproducible steps from data intake through model evaluation.

By accurately predicting which passengers were transported, this project aims to assist rescue operations in efficiently locating and retrieving lost travelers, potentially saving thousands of lives.

*Let's begin the analysis to uncover the mysteries hidden in the Spaceship Titanic's data.*

In [1]:
import sys
sys.path.append("../src")

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.impute import SimpleImputer
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.feature_selection import SelectKBest, f_classif
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier
import optuna
from spaceship_titanic_utils import *
warnings.filterwarnings("ignore")
from IPython.display import Image

/Users/vytautasbunevicius/kaggle-spaceship-titanic/venv/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [3]:
train_data = pd.read_csv("../data/train.csv")
test_data = pd.read_csv("../data/test.csv")

In [4]:
print("Train dataset first few rows:")
print(train_data.head())
print("\nTrain dataset shape:", train_data.shape)

Train dataset first few rows:
  PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0001_01     Europa     False  B/0/P  TRAPPIST-1e  39.0  False   
1     0002_01      Earth     False  F/0/S  TRAPPIST-1e  24.0  False   
2     0003_01     Europa     False  A/0/S  TRAPPIST-1e  58.0   True   
3     0003_02     Europa     False  A/0/S  TRAPPIST-1e  33.0  False   
4     0004_01      Earth     False  F/1/S  TRAPPIST-1e  16.0  False   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0          0.0        0.0           0.0     0.0     0.0    Maham Ofracculy   
1        109.0        9.0          25.0   549.0    44.0       Juanna Vines   
2         43.0     3576.0           0.0  6715.0    49.0      Altark Susent   
3          0.0     1283.0         371.0  3329.0   193.0       Solam Susent   
4        303.0       70.0         151.0   565.0     2.0  Willy Santantines   

   Transported  
0        False  
1         True  
2        False  
3     

In [5]:
print("\nTest dataset first few rows:")
print(test_data.head())
print("\nTest dataset shape:", test_data.shape)


Test dataset first few rows:
  PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0013_01      Earth      True  G/3/S  TRAPPIST-1e  27.0  False   
1     0018_01      Earth     False  F/4/S  TRAPPIST-1e  19.0  False   
2     0019_01     Europa      True  C/0/S  55 Cancri e  31.0  False   
3     0021_01     Europa     False  C/1/S  TRAPPIST-1e  38.0  False   
4     0023_01      Earth     False  F/5/S  TRAPPIST-1e  20.0  False   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck              Name  
0          0.0        0.0           0.0     0.0     0.0   Nelly Carsoning  
1          0.0        9.0           0.0  2823.0     0.0    Lerome Peckers  
2          0.0        0.0           0.0     0.0     0.0   Sabih Unhearfus  
3          0.0     6652.0           0.0   181.0   585.0  Meratz Caltilter  
4         10.0        0.0         635.0     0.0     0.0   Brence Harperez  

Test dataset shape: (4277, 13)


Data loaded successfully. Train set has 8693 samples and 14 features, while test set has 4277 samples and 13 features (excluding the target variable).

Now, let's display basic information about the dataset itself.

In [6]:
print("\nTraining data info:")
print(train_data.info())


Training data info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8693 entries, 0 to 8692
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   PassengerId   8693 non-null   object 
 1   HomePlanet    8492 non-null   object 
 2   CryoSleep     8476 non-null   object 
 3   Cabin         8494 non-null   object 
 4   Destination   8511 non-null   object 
 5   Age           8514 non-null   float64
 6   VIP           8490 non-null   object 
 7   RoomService   8512 non-null   float64
 8   FoodCourt     8510 non-null   float64
 9   ShoppingMall  8485 non-null   float64
 10  Spa           8510 non-null   float64
 11  VRDeck        8505 non-null   float64
 12  Name          8493 non-null   object 
 13  Transported   8693 non-null   bool   
dtypes: bool(1), float64(6), object(7)
memory usage: 891.5+ KB
None


The dataset contains a mix of numeric and categorical variables. Some columns have missing values.

Moving forward to checking for missing values.

In [7]:
missing_percentages = (train_data.isnull().sum() / len(train_data) * 100).sort_values(ascending=False)
print("Missing value percentages:")
print(missing_percentages[missing_percentages > 0])

Missing value percentages:
CryoSleep       2.496261
ShoppingMall    2.392730
VIP             2.335212
HomePlanet      2.312205
Name            2.300702
Cabin           2.289198
VRDeck          2.162660
FoodCourt       2.105142
Spa             2.105142
Destination     2.093639
RoomService     2.082135
Age             2.059128
dtype: float64


Most features have about 2-2.5% missing values. Let's handle these missing values by removing rows with any missing data.

In [8]:
train_data_cleaned = handle_missing_values(train_data)
test_data_cleaned = handle_missing_values(test_data)

Columns dropped due to >30.0% missing values: []
Rows removed due to missing values: 2087
Columns dropped due to >30.0% missing values: []
Rows removed due to missing values: 996


In [9]:
print("Train dataset shape after handling missing values:", train_data_cleaned.shape)
print("Test dataset shape after handling missing values:", test_data_cleaned.shape)

Train dataset shape after handling missing values: (6606, 14)
Test dataset shape after handling missing values: (3281, 13)


We've removed rows with missing values. The train dataset shape is now (6606, 14) and the test dataset shape is (3281, 13). This approach maintains data integrity without introducing potential biases from imputation.

Next, let's check for duplicate rows in the dataset to ensure data quality.

In [10]:
duplicates = train_data_cleaned.duplicated().sum()
print(f"\nNumber of duplicate rows: {duplicates}")
if duplicates > 0:
    train_data_cleaned = train_data_cleaned.drop_duplicates().reset_index(drop=True)
    print("Duplicates removed. New training set shape:", train_data_cleaned.shape)


Number of duplicate rows: 0


We found 0 duplicate rows in the dataset. This indicates good data quality in terms of uniqueness of records.

Now, let's display basic summary statistics for our dataset to get an overview of the distribution of our numerical features."

In [11]:
print("\nSummary statistics:\n")
print(train_data_cleaned.describe().T)


Summary statistics:

               count        mean          std  min   25%   50%    75%      max
Age           6606.0   28.894036    14.533429  0.0  19.0  27.0  38.00     79.0
RoomService   6606.0  222.991674   644.987936  0.0   0.0   0.0  49.00   9920.0
FoodCourt     6606.0  478.958523  1678.592291  0.0   0.0   0.0  82.75  29813.0
ShoppingMall  6606.0  178.356494   576.328407  0.0   0.0   0.0  30.00  12253.0
Spa           6606.0  313.161520  1144.016291  0.0   0.0   0.0  65.00  22408.0
VRDeck        6606.0  303.780048  1127.142166  0.0   0.0   0.0  52.00  20336.0


The summary statistics reveal interesting patterns in our numerical features:

**Age:** Ranges from 0 to 79 years, with a mean of about 28.9 years. The median (50%) is 27 years, suggesting a relatively young passenger population.

**Spending features (RoomService, FoodCourt, ShoppingMall, Spa, VRDeck):**

All have minimums of 0 and high maximum values, indicating wide ranges in spending. The means are much higher than the medians (which are all 0), suggesting highly skewed distributions. Large standard deviations compared to means indicate high variability in spending patterns.

**These statistics suggest we may need to consider log transformations or other techniques to handle the skewed spending data in our analysis. FoodCourt shows the highest mean spending and maximum value among the features.**

Next up, we can move towards checking the distributions of our numerical features using histograms.

**Note: We're not including 'Cabin', 'PassengerId', and 'Name' in our initial distribution analysis for the following reasons:**

1. Cabin: This feature contains multiple pieces of information (deck, cabin number, side) and would benefit from feature engineering before analysis. We'll address this in our feature engineering step.

2. PassengerId: This is a unique identifier for each passenger or group and doesn't provide direct information about the likelihood of being transported. However, we may extract useful information from it later, such as group size or travel companions.

3. Name: While names could potentially provide information about family groups or cultural background, they require more complex text analysis techniques. We'll consider whether to use this information in our feature engineering step.

In [15]:
numerical_features = ['Age', 'RoomService', 'FoodCourt', 'ShoppingMall', 'Spa', 'VRDeck']
numerical_features_one = ['Age', 'RoomService', 'FoodCourt']
numerical_features_two = ['ShoppingMall', 'Spa', 'VRDeck']

plot_combined_histograms(
    train_data_cleaned,
    numerical_features_one,
    nbins=30,
    save_path="../images/numerical_features_one.png",
)

1. **Age:** Roughly normal distribution, peaking around 20-30 years. Slight right skew with fewer older passengers.
2. **RoomService:** Extremely right-skewed. Most passengers use little to no room service, with a long tail of high spenders.
3. **FoodCourt:** Similar to RoomService, highly right-skewed. Majority spend little, but some passengers have very high food court expenses.

In [16]:
plot_combined_histograms(
    train_data_cleaned,
    numerical_features_two,
    nbins=30,
    save_path="../images/numerical_features_two.png",
)

1. **ShoppingMall:** Heavily right-skewed. Most passengers spend little or nothing, with a few high spenders.
2. **Spa:** Nearly identical pattern to ShoppingMall. Majority low/no spending, long tail of high spenders.
3. **VRDeck:** Also right-skewed, but slightly less extreme. More passengers use VRDeck compared to other amenities, still with some very high spenders.

Next up, let's move on to checking the distributions of our categorical features using bar plots.

In [17]:
categorical_features = ['HomePlanet', 'CryoSleep', 'Destination', 'VIP', 'Transported']
categorical_features_one = ['HomePlanet', 'CryoSleep', 'Destination']
categorical_features_two = ['VIP', 'Transported']

In [24]:
plot_combined_bar_charts(
    train_data_cleaned,
    categorical_features_one,
    save_path="../images/categorical_distributions_one.png",
)

1. **HomePlanet:** Earth is the most common origin, followed by Europa, then Mars. Clear imbalance between planets.
2. **CryoSleep:** More passengers are not in cryosleep (false) than in cryosleep (true). Roughly 2:1 ratio.
3. **Destination:** TRAPPIST-1e is the most popular destination by far, followed by 55 Cancri e, then PSO J318.5-22.

In [26]:
plot_combined_bar_charts(
    train_data_cleaned,
    categorical_features_two,
    save_path="../images/categorical_distributions_two.png",
)

1. **VIP:** Vast majority of passengers are not VIP. Very small proportion of VIP passengers.
2. **Transported:** Nearly balanced distribution between transported (true) and not transported (false) passengers, with slightly more transported.

In [21]:
%run -i ../src/spaceship_titanic_utils.py